### ISSUE TITLE
Detect redundant conditions in an `if` chain

### ISSUE THREAD

**Issue by jfmengels** [jfmengels/elm-review-simplify#22]
**What would you like to see:**

Report unreachable branches and remove them and simplifiable conditions

Prior art: https://eslint.org/docs/latest/rules/no-dupe-else-if

**What problems does it solve:**

This could report copy-pasted branches 

**Example of things the rule would report:**

1. Duplicate conditions in nested if expressions

```diff
 if a then
     1

 else if b then
     2

-else if b then
-    3

 else
     4
```

2. Simplifiable conditions because the conditions will always evaluate to false in the `else` chain

```diff
 if a then
     1

-else if a || b then
+else if b then
     2

 else
     3
```

3. Simplifiable conditions based on prior knowledge
```diff
 if a then
     1

 else if b then
     -- we know that `a` is False here
-    if a then
-        2
- 
-    else
         3
 
 else
     4
```

**Example of things the rule would not report:**

```elm
if a then
     1

else if b then
     2

else
     3
```

**I am looking for:**

Someone to add this check to the `Simplify` rule :slightly_smiling_face: 

Not everything in this proposal needs to be added at once, incremental improvements are fine too :+1: 

This is a way I can imagine `1.` being implemented: Whenever we find an if expression, we crawl the body of the `else` expressions to gather the different conditions. We then compare them (using `Simplify.Normalize.compare`) to find out whether they are equal or not.
For `2.` and `3.` we probably need a tiny bit of a different approach, I haven't put too much thought into yet.

To keep things simple, we can also replace the conditions to `True`/`False`, as the rule will simplify that further on its own already. Doing it directly would make the fixes faster though.

Please reply if you're interested in working on this! If you do, please write lots of tests :)

**Comment by sebsheep** [comment]
Hey it looks really fun and I'd be very interested working on this !!!

Unfortunately, I have a bunch of things to do before leaving for summer break in less than a week....

**Comment by sebsheep** [comment]
other possible approach:
on every `if`, store the expression in a `Dict BoolExpr Bool` "context" and every time we have such a `BoolExpr` that is used, propose to simplify it to `True` or `False`. Similar crawling for the `else` branch.

**Comment by gacallea** [comment]
Going to attempt this. Can't promise. I'm glad for the additional comments, as they match my idea of tackling this and answer a few questions already. :)

**Comment by gacallea** [comment]
I begrudgingly realized that his is above my current skills, and that I won't be able to put work on this until I have improved them. If anyone wants to take over. Martin Janizczek has implemented a proof of concept solution that can be adapted. It is on Ellie, shared in this thread: https://twitter.com/jfmengels/status/1544042722738274309

Allow me to close on a positive note. The comforting bits are: 1) it felt good to read 6KLOC for the very first time and kinda understanding what they generally did. 2) having the right idea (e.g: being in the ballpark) about the solution. I guess I lacked experience and more knowledge under my belt.

Cheers
